qualtran/bloqs/arithmetic/multiplication.py

```diff
@@ -12,16 +12,78 @@
 #  See the License for the specific language governing permissions and
 #  limitations under the License.
 
-from typing import Set, TYPE_CHECKING
+from typing import Any, Dict, Iterable, Sequence, Set, TYPE_CHECKING, Union
 
+import cirq
 from attrs import frozen
 
-from qualtran import Bloq, Register, Side, Signature
+from qualtran import Bloq, GateWithRegisters, Register, Side, Signature
 from qualtran.bloqs.basic_gates import Toffoli
 from qualtran.cirq_interop.t_complexity_protocol import TComplexity
 
 if TYPE_CHECKING:
     from qualtran.resource_counting import BloqCountT, SympySymbolAllocator
+    from qualtran.simulation.classical_sim import ClassicalValT
+
+
+@frozen
+class PlusEqualProduct(GateWithRegisters, cirq.ArithmeticGate):
+    """Performs result += a * b"""
+
+    a_bitsize: int
+    b_bitsize: int
+    result_bitsize: int
+    adjoint: bool = False
+
+    def short_name(self) -> str:
+        return "result -= a*b" if self.adjoint else "result += a*b"
+
+    @property
+    def signature(self) -> 'Signature':
+        return Signature.build(a=self.a_bitsize, b=self.b_bitsize, result=self.result_bitsize)
+
+    def registers(self) -> Sequence[Union[int, Sequence[int]]]:
+        return [2] * self.a_bitsize, [2] * self.b_bitsize, [2] * self.result_bitsize
+
+    def apply(self, a: int, b: int, result: int) -> Union[int, Iterable[int]]:
+        return a, b, (result + a * b * ((-1) ** self.adjoint)) % (2**self.result_bitsize)
+
+    def with_registers(self, *new_registers: Union[int, Sequence[int]]):
+        raise NotImplementedError("Not needed.")
+
+    def on_classical_vals(self, *args) -> Dict[str, 'ClassicalValT']:
+        return dict(zip([reg.name for reg in self.signature], self.apply(*args)))
+
+    def _t_complexity_(self) -> 'TComplexity':
+        return TComplexity(t=4 * max(self.a_bitsize, self.b_bitsize) ** 2)
```

Comment:
Naively I would have thought this would cost 2 n^2 - 1 Toffolis right, (multiplication is 2 n^2 - n) and addition is n - 1), wouldn't this make 8 max(n_a, n_b)^2 Ts (not 4) if we ignore the constants.

Reply:
I was also ignoring the constant `2` (i.e. I just considered O(n ** 2)). Updated the T-complexity and added a comment.